helix-core/src/test/java/org/apache/helix/controller/rebalancer/constraint/TestAbnormalStatesResolverMonitor.java

```diff
@@ -0,0 +1,89 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.lang.management.ManagementFactory;
+import java.util.Random;
+import javax.management.AttributeNotFoundException;
+import javax.management.InstanceNotFoundException;
+import javax.management.MBeanException;
+import javax.management.MBeanServer;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+
+import org.apache.helix.model.MasterSlaveSMD;
+import org.apache.helix.monitoring.mbeans.MonitorDomainNames;
+import org.apache.helix.monitoring.metrics.AbnormalStatesMetricCollector;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class TestAbnormalStatesResolverMonitor {
+  private static final MBeanServer beanServer = ManagementFactory.getPlatformMBeanServer();
+
+  @Test
+  public void testMonitorResolver()
+      throws MalformedObjectNameException, AttributeNotFoundException, MBeanException,
+      ReflectionException, InstanceNotFoundException {
+    final String clusterName = "TestCluster";
+
+    final String TEST_RESOLVER_MONITOR_MBEAN_NAME = String
+        .format("%s:%s=%s, %s=%s.%s", MonitorDomainNames.Rebalancer, "ClusterName", clusterName,
+            "EntityName", "AbnormalStates", MasterSlaveSMD.name);
+    final ObjectName TEST_RESOLVER_MONITOR_MBEAN_OBJECT_NAME =
+        new ObjectName(TEST_RESOLVER_MONITOR_MBEAN_NAME);
+
+    Assert.assertFalse(beanServer.isRegistered(TEST_RESOLVER_MONITOR_MBEAN_OBJECT_NAME));
+
+    // Update the resolver configuration for MasterSlave state model.
+    MonitoredAbnormalResolver monitoredAbnormalResolver =
+        new MonitoredAbnormalResolver(new MockAbnormalStateResolver(), clusterName,
+            MasterSlaveSMD.name);
+
+    // Validate if the MBean has been registered
+    Assert.assertTrue(beanServer.isRegistered(TEST_RESOLVER_MONITOR_MBEAN_OBJECT_NAME));
+    Assert.assertEquals(beanServer.getAttribute(TEST_RESOLVER_MONITOR_MBEAN_OBJECT_NAME,
+        AbnormalStatesMetricCollector.AbnormalStatesMetricNames.AbnormalStatePartitionCounter.name()),
+        0L);
+    Assert.assertEquals(beanServer.getAttribute(TEST_RESOLVER_MONITOR_MBEAN_OBJECT_NAME,
+        AbnormalStatesMetricCollector.AbnormalStatesMetricNames.RecoveryAttemptCounter.name()), 0L);
+    // Validate if the metrics recording methods work as expected
+    Random ran = new Random(System.currentTimeMillis());
+    Long expectation = 1L + ran.nextInt(10);
```

Comment:
`long`?

Reply:
Object is required to use assertEquals.

Reply:
Did you mean `beanServer.getAttribute()` returns an Object and expectation needs a Long object as well?
```
Assert.assertEquals(beanServer.getAttribute(TEST_RESOLVER_MONITOR_MBEAN_OBJECT_NAME,
    AbnormalStatesMetricCollector.AbnormalStatesMetricNames.AbnormalStatePartitionCounter.name()),
        expectation);
```

I think primitives work. And confirmed in your patch changing to long works.

Reply:
I'd rather introduce less hidden type convert in the test logic. Moreover, this is a test. We don't really care about the tiny performance difference. We care about test stability more than the other facts.

Reply:
Actually, I believe the check will fail if using primitive types. The Assert thinks the values are different.

Reply:
1. Regarding object type, I understand you would like to make sure `Assert.assertEquals(Object, Object)` is called, instead of auto boxing long type. I am totally fine here and wouldn't care about performance.

> Actually, I believe the check will fail if using primitive types. The Assert thinks the values are different.
The check would not fail. As I said, it would automatically auto box the primitive value to a boxed primitive (Long). So I don't think (and run it without failure) the check would fail.

Reply:
We could close this discussion if you don't have different opinions :)